### DISCUSSION TITLE
Simpler path bar

### DISCUSSION THREAD

**Discussion by tintou** [elementary/files#2166] [Ideas]
### Problem

The current code for the path bar is really non-obvious and will be difficult to move to GTK4

### Proposal

Have a less complicated path bar.
In macOS the path bar is even hidden by default, while I don't think that it is what we want here, having it styled with simple arrows would make it simpler and easier to maintain.

### Prior Art (Optional)

Finder: ![image](https://user-images.githubusercontent.com/745535/164322333-55091f13-7f81-411e-a8ab-ce4e10ee98f6.png)

**Comment by danirabbit** [comment]
Yeah I think the slash style like in GNOME Files Gtk4 port looks nice too

![image](https://user-images.githubusercontent.com/7277719/164332790-63192f0b-daf6-482f-9441-3b70c4b9b90d.jpeg)

**Reply by danirabbit** [reply]
Specifically regarding the pathbar style, here's a mockup using slashes for the separator instead of the chevrons we're using now. I think this is pretty clean looking

![files-path](https://user-images.githubusercontent.com/7277719/229929080-38db385a-aa5b-4e68-956a-4270a28e22c8.png)

**Comment by jeremypw** [comment]
Also need to decide how to incorporate the Searchbar.

**Reply by danirabbit** [reply]
I'd like to keep the searchbar and pathbar as a single widget if possible. I think we've had a pretty long precedent of this not just in files, but it's how the pathbar has worked for every web browser for many years now.

This could be broken up into a series of issue reports related to search, but I have a mockup here:

![files-search-ico](https://user-images.githubusercontent.com/7277719/229930141-1ac0020e-dbf4-4ceb-bc2f-d0a2343ce3b0.png)
* This mockup assumes a search taking place in "Home/Projects"
* I think one of the pain points that we have right now with the current pathbar is that the top result should always be the "type ahead" result and not something else. I think this is a much more predictable behavior and makes keyboard navigation a lot faster.
* Instead of saying "In This Folder", use the name of the current folder since we lose that information while searching
* Instead of having a "More results" item in each section that continually grows the search window, I think we should have an advanced search page that is like a virtual folder view where we can show all of the returned results and have filters etc

**Reply by danirabbit** [reply]
Another kind of search exploration: Sometimes a result appears under multiple headers, so instead of sorting the individual headers we could give a weight for each one and then sort results individually according that weight. We could then get rid of the headers altogether and use icons to give context to a result. 
![files-search-bigresults](https://user-images.githubusercontent.com/7277719/229936364-274e5f14-22b4-4fd9-80ce-6603dd9db1b8.png)

**Comment by hanaral** [comment]
@jeremypw Would it be _too_ wrong to have a search bar like appcenter or music? If there isn't going to be a text based pathbar then the next most obvious thing is a permanent one.

**Comment by jeremypw** [comment]
It would certainly improve discoverability to have it permanently visible although it may increase the minimum width. I guess it need not be too large as search terms are generally short. Other possibilities would be a Code-like search bar in a revealer, or having it in a Stack with the path bar.  Anyway it would be a separate widget.

**Comment by PlatonB** [comment]
> Yeah I think the slash style like in GNOME Files Gtk4 port looks nice too

Pathbar in Gnome seems bad to me because there is no way to copy the current directory path.

**Comment by jeremypw** [comment]
@danirabbit   Thanks for the fresh ideas on the design. It should be possible to continue to have the navigation and search entries being the same entry but I would like to separate out the breadcrumbs into a separate widget rather than custom drawing over the entry as it simplifies the code considerably. The widgets can be switched with a stack as required.

Are these changes something to be done in the Gtk3 master version or brought in with the Gtk4 version?

**Reply by tintou** [reply]
It depends on how far the GTK4 port is already, if it helps to see the end result, you might want to use the GTK3 version as the implementation will probably be similar in GTK4

**Reply by jeremypw** [reply]
The draft Gtk4 port is fairly complete apart from the search bar, styling issues, a couple of plugins and testing for RTL and corner case regressions.  I have taken the approach of an extensive rewrite using (I hope) undeprecated Gtk4 widgets as much as possible and making some simplifications so the code is about 9000 lines shorter (some of that will be used up by the searchbar though).   Some of what I have done can be backported to the Gtk3 version or delayed to reduce the diff considerably. Probably the only realistic way to get it merged into master.  Alternative is to create a new development focus in main and treat it as a new starting point I guess, like Music.

**Reply by jeremypw** [reply]
I agree if we are completely rewriting the searchbar then I could do it in Gtk3 keeping in mind the need to port it to Gtk4 as easily as possible.  I have a draft version of SearchResults semi-working in Gtk4 which I havent pushed yet.  A lot of stuff had to be changed as it uses complicated device grabbing on a window which cannot be done in Gtk4.  I have tried using a standard Popover instead and just put the existing TreeView into it.  It seems to work.

**Reply by tintou** [reply]
Okay, then make it a GTK4 exclusive :smile: no problem for that
